src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -399,6 +399,25 @@ static inline bool target_needs_far_branch(address addr) {
   return !CodeCache::is_non_nmethod(addr);
 }
 
+static inline bool target_needs_trampoline(Address entry) {
+  if (entry.rspec().type() == relocInfo::runtime_call_type) {
+    // Runtime calls are statically bound.
+    // Once they are generated neither a caller nor a callee address cannot be changed.
+    // Check whether a far branch is needed to reach the target.
+    return target_needs_far_branch(entry.target());
```

Comment:
From what I understand `target_needs_far_branch()` works only for stubs which are inside CodeCache.
`runtime` method, as you correctly pointed, is code in VM which can be in any distance from CodeCache.
For example, it will give incorrect answer if CodeCache size is small:
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L391

Or I am confused by the name of this method `target_needs_trampoline` and the comment.
Do you actually mean `need_far_branch_for_trampoline`?

Reply:
> runtime method, as you correctly pointed, is code in VM which can be in any distance from CodeCache.

Good catch. The function works with assumptions of `MacroAssembler::trampoline_call`. 
`MacroAssembler::trampoline_call` implicitly assumes the target is inside `CodeCache`:
```
// Maybe emit a call via a trampoline.  If the code cache is small
// trampolines won't be emitted.
```
I'll move the code of `target_needs_far_branch` inside of `MacroAssembler::trampoline_call`.
I'll also add an assert checking the target address is inside `CodeCache`.

Reply:
I suggest to put somewhere description where `far` calls and `trampoline` should/can be used.

Looking on `MacroAssembler::far_call()` and based on the assert there it targets only entries inside CodeCache. Which is strange because if `runtime` entry inside the 2gb range from any place in CodeCache you can use it. We have such check for x86: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/assembler_x86.cpp#L11701
And you don't need trampoline if you can use `far` call to such `runtime` entry.

If you need trampoline there is also choice how you call/jump to **it**. Trampoline code is inside CodeCache and `target_needs_far_branch()` works to find type of instruction to reach *trampoline*.  That is what I though  your changes are doing.

Reply:
Yes, you are right. There is an issue that for runtime calls into `CodeCaceh` both `far_call` and `trampoline_call` can be used.
`MacroAssembler::trampoline_call` should be used for runtime calls outside `CodeCache` and Java calls.  If a runtime call is into `CodeCache` then `far_call` should be used.